### PR TITLE
Upgrade ACE Editor from 1.2.2 to 1.15.0

### PR DESCRIPTION
**Pull request by basil**
Upgrade [ACE Editor](https://ace.c9.io/) from 1.2.2 (released on October 28, 2015) to 1.15.0 (released on January 25, 2023).

### Implementation

Followed the example in the [Webpack demo here](https://github.com/ajaxorg/ace/blob/159aa70d551530f2866fb0006fad37bd75e60dda/demo/webpack/demo.js).

### Testing done

- Create a new Pipeline job and try samples
- Autocomplete Pipeline-specific snippets
- Introduce a syntax error and verify validation fails
- Fix a syntax error and verify validation succeeds
- Resize the editor

### CODE REVIEW

File: plugin/package.json [hunk 1]
```diff
@@ -39,7 +39,7 @@
     "webpack-fix-style-only-entries": "^0.6.1"
   },
   "dependencies": {
-    "jenkins-js-modules": "1.3.0",
+    "ace-builds": "^1.15.0",
```
Comment (basil) [inline plugin/package.json:42]: With the removal of the dependency on [this plugin](https://plugins.jenkins.io/ace-editor/), `jenkins-js-modules` is no longer needed.

File: plugin/pom.xml [hunk 2]
```diff
@@ -42,8 +42,8 @@
     </licenses>
     <properties>
         <no-test-jar>false</no-test-jar>
-        <node.version>16.17.0</node.version>
-        <npm.version>8.18.0</npm.version>
+        <node.version>18.13.0</node.version>
+        <npm.version>8.19.3</npm.version>
```
Comment (basil) [inline plugin/pom.xml:46]: Upgrading these to recent versions to take advantage of the existing testing cycle.

File: plugin/pom.xml [hunk 3]
```diff
@@ -165,16 +165,6 @@
...
-        <dependency>
-            <groupId>org.jenkins-ci.ui</groupId>
-            <artifactId>ace-editor</artifactId>
-            <exclusions>
-                <exclusion> <!-- TODO remove when ace-editor is refreshed -->
-                    <groupId>javax.servlet</groupId>
-                    <artifactId>servlet-api</artifactId>
-                </exclusion>
-            </exclusions>
-        </dependency>
```
Comment (basil) [inline plugin/pom.xml:177]: Bye-bye https://github.com/jenkinsci/js-libs

File: plugin/src/main/js/workflow-editor.js [hunk 4]
```diff
@@ -30,24 +32,18 @@ jenkinsJSModules.import('ace-editor:ace-editor-122')
...
+                // https://stackoverflow.com/a/66923593
+                var snippetManager = ace.require('ace/snippets').snippetManager;
+                var snippetContent = ace.require('ace/snippets/groovy').snippetText;
+                var snippets = snippetManager.parseSnippetFile(snippetContent);
+                snippetManager.register(snippets, 'groovy');
 
-                acePack.addScript('ext-language_tools.js', function() {
-                    ace.require("ace/ext/language_tools");
-
-                    editor.$blockScrolling = Infinity;
```
Comment (basil) [inline plugin/src/main/js/workflow-editor.js:50]: Removed without replacement in https://github.com/ajaxorg/ace/pull/2966.

File: plugin/src/main/js/workflow-editor.js [hunk 5]
```diff
@@ -60,7 +56,7 @@ jenkinsJSModules.import('ace-editor:ace-editor-122')
                     });
 
                     editor.setValue(textarea.val(), 1);
-                    editor.getSession().on('change', function() {
+                    editor.getSession().on('change', function(delta) {
```
Comment (basil) [inline plugin/src/main/js/workflow-editor.js:59]: See "Listening to Events" on [this page](https://ace.c9.io/#nav=howto).

**Review by jglick — APPROVED** [review]

**Comment by timja** [comment]
Likely caused https://issues.jenkins.io/browse/JENKINS-70591

**Comment by basil** [comment]
> Likely caused https://issues.jenkins.io/browse/JENKINS-70591

What makes us think the problem was likely caused by this PR rather than #656? Have we done any bisection or are we making assertions without evidence?

**Comment by timja** [comment]
I have not bisected. Which is why I said likely.

**Pull request merged**